### ISSUE TITLE
conda-build 3.0.0alpha3 produces conda packages with new build strings

### ISSUE THREAD

**Issue by ilanschnell** [conda/conda-build#1919]
I'm raising some concerns about the new build strings which conda-build 3.0 is creating:
  * the new build strings are different from the existing ones, and this might cause problems for existing users.  The build string does not need to be of a specific format, as far as conda is concerned.  The existing convention, e.g. `np19py27_1` has existed since the beginning of the conda project (this convention actually predates conda-build) and many people are probably very used to it by now.  Moreover, people (such as myself, but likey others as well) have written scripts around conda-build which rely on the existing build string naming convention.
  * the new build strings contain opaque hashes, which makes users wonder what exactly they mean and which information they contain.  In creating the conda ecosystem, one of it's core design principles has always been simplicity and transparency.  The hashes are not transparent and therefore violate this design principle.

**Comment by msarahan** [comment]
We have discussed both of these issues at length internally.  Please don't waste time repeating yourself.

1. Sorry, but everyone agrees that filenames as API is far too limiting.  Conda no longer uses this convention.  I'd be happy to help you adapt your tools.  This is a major version bump, and yes, I do expect it will break some people.  I hope that the expanded capabilities make it worth any effort to fix tools.  If you would embrace the new capability, maybe you'd see more of the equation.

2. This is exactly what the ``conda inspect hash-inputs`` command was written for.  Granted, it is not quite as immediately visible as information in filenames.

This change was necessary to generalize the pinning concept.  Your ``x.x`` idea for numpy simply can't scale to all of the things that people want to pin in the same way that numpy is pinned.  If you want me to get rid of hashes, you need to propose an alternative that is at least as good in terms of supporting arbitrary pinning.

**Comment by ilanschnell** [comment]
I'm not suggesting to remove the new hashes.  There are no doubt situations where the new hashes are necessary.  However, there are also situations in which people are going to rely on the old build sting convention.  In order to make it possible for these people to use conda-build 3.0, it would be helpful to provide an "--old-build-string" option.  Otherwise, these people will simply stay on conda-build 2.x, and conda-build 3.x will get less traction.  Whenever, we break backwards compatibility we need to worry about a transition plan, and about maintaining the old version.

**Comment by teoliphant** [comment]
Ilan brings up a good point that couldn't conda-build 3 have an option to allow a different build string to be used in cases where someone did not want to use the hash-in-the-filename convention.     Is there a way to override the build-string? e.g. " --override-build-string"

**Comment by msarahan** [comment]
If you do that, then you need another way of saying what your whole ecosystem supports.  How do you encode glibc version support?  Do you jam everything into the filename?

I can add the flag to disable it, but I think it is only safe to use if the package system is self-consistent regarding CDTs.  This has historically been the case, but it is getting harder to support.  We will have to switch to the new C++11 ABI for Linux at some point, and it is absolutely necessary to have a way to differentiate between packages built with one ABI or the other.

Please show me where we break backwards compatibility - point to code.  Our two choices here are to fix that code, or to continue sitting on the time bomb of conflicting package names, because the current names are simply incapable of capturing sufficient complexity.  I have offered to fix any broken code before.  I'm still happy to do that.  I just don't want to choose the time bomb on account of a gut feeling.

**Comment by teoliphant** [comment]
No, you just need a way of overriding the naming convention.    It's not really a matter of backward incompatibility, it's a matter of letting people still decide whether they want the build-string to convey this information or not.     Those who don't use this hash mechanism will have to come up with some other differentiating mechanism.   It let's us move forward with this feature in conda-build without also solving the problem of pushing everyone into the direction of using this feature right out of the gate.

**Comment by teoliphant** [comment]
As mentioned in the #2010 thread, a --build-string option that can be used to just pass in whatever approach the package builder wants to the name of the file, might be preferable to --old-build-string.  Then, conda-build can actually drop the incomplete logic that has been used in the past to pin to only python and numpy.

**Comment by msarahan** [comment]
I'm actually not in favor of totally dropping the old way.  I actually fixed several issues with it between conda-build 2 and 3.  I think it captures the "principal axes" if you will, and it is helpful at a glance.  I think that keeping it in combination with the hashes is the best way forward.

The only thing at issue here is what goes into the filename - not what is pinned.  That's much different logic, and if that has gotten incompatible in any way, that's a major issue.  I worked very hard to keep that compatible.
